File: esbmc.cpp/validador_preco/validador_valor_positivo.cpp
#include <cstdlib>

#define MAX_PRICE_LEN 6

using namespace std;

bool validate_positive_value(const char price[MAX_PRICE_LEN], int len) {
    double value = 0;
    double factor = 1;
    bool after_dot = false;
    int decimals = 0;

    for (int i = 0; i < len; i++) {
        char c = price[i];
        if (c == '.') {
            after_dot = true;
            continue;
        }
        value = value * 10 + (c - '0');
        if (after_dot) {
            factor *= 10;
            decimals++;
        }
    }

    value /= factor;

    __ESBMC_assert(value > 0, "O preço deve ser um valor positivo.");
    return true;
}

int main() {
    char price[MAX_PRICE_LEN];
    int len;

    __ESBMC_assume(len >= 1 && len <= 5);

    validate_positive_value(price, len);

    return 0;
}

File: esbmc.py/Testes02/validador_preco/validador_valor_positivo.py
MAX_PRICE_LEN: int = 6

def validate_positive_value(price: str, length: int) -> bool:
    value: float = 0.0
    factor: float = 1.0
    after_dot: bool = False
    decimals: int = 0
    i: int = 0

    while i < length:
        c: str = price[i]
        if c == '.':
            after_dot = True
            i += 1
            continue

        value = value * 10 + (ord(c) - ord('0'))

        if after_dot:
            factor *= 10
            decimals += 1

        i += 1

    value /= factor

    assert value > 0, "O preço deve ser um valor positivo."
    return True


def main() -> None:
    price: str = input()        # entrada simbólica
    length: int = len(price)

    # equivalente ao __ESBMC_assume
    assert 1 <= length <= 5

    validate_positive_value(price, length)


main()
